internal/provider/resource_secret.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the Apache License, Version 2.0, see LICENCE file for details.
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/juju/terraform-provider-juju/internal/juju"
+)
+
+// Ensure provider defined types fully satisfy framework interfaces.
+var _ resource.Resource = &secretResource{}
+var _ resource.ResourceWithConfigure = &secretResource{}
+var _ resource.ResourceWithImportState = &secretResource{}
+
+func NewSecretResource() resource.Resource {
+	return &secretResource{}
+}
+
+type secretResource struct {
+	client *juju.Client
+
+	// subCtx is the context created with the new tflog subsystem for applications.
+	subCtx context.Context
+}
+
+type secretResourceModel struct {
+	// Model to which the secret belongs. This attribute is required for all actions.
+	Model types.String `tfsdk:"model"`
+	// Name of the secret to be updated or removed. This attribute is required for 'update' and 'remove' actions.
+	Name types.String `tfsdk:"name"`
+	// Value of the secret to be added or updated. This attribute is required for 'add' and 'update' actions.
+	// Template: [<key>[#base64]]=<value>[ ...]
+	Value types.Map `tfsdk:"value"`
+	// SecretId is the ID of the secret to be updated or removed. This attribute is required for 'update' and 'remove' actions.
+	SecretId types.String `tfsdk:"secret_id"`
+	// Info is the description of the secret. This attribute is optional for all actions.
+	Info types.String `tfsdk:"info"`
+}
+
+func (s *secretResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+func (s *secretResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_secret"
+}
+
+func (s *secretResource) Schema(_ context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "A resource that represents a Juju secret.",
+		Attributes: map[string]schema.Attribute{
+			"model": schema.StringAttribute{
+				Description: "The model in which the secret belongs.",
+				Required:    true,
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of the secret. Note that the name cannot be changed once set.",
```

Comment:
s/The name of the secret. Note that the name cannot be changed once set/The name of the secret/